README.md
# With Loader

A Flutter package that provides loading functionality to a widget. Just Warp your widget with `WithLoader` and give all the necessary parameters.

## Features

* Can be helpful when you need a `Future` function to be performed `onTap` of a Button.
* You can customise the `child` opacity while it is in loading state.
* It exposes the result of the `Future` function, through which you can decide the `success` or `failure` condition of that function. 
* According to this `successCondition` it will perform `onSuccess` or `onFailure`.

## Getting started

Add `with_loader` to `pubspec.yaml` of your project:

```yaml
dependencies:
  with_loader: ^0.0.1
```

Import it in your Dart code:

```dart
import 'package:with_loader/with_loader.dart';
```

## Usage

Wrap any widget with `WithLoader<T>()`:  

```dart
WithLoader<Type>(
  child: AnyWidget()
),
```

Some required `parameters` to make this widget work:

```dart
WithLoader<String>(
  future: () => Future.delayed(const Duration(seconds: 2), () => "Success"),
  successCondition: (result) => result == "Success",
  child: Container(
      decoration: BoxDecoration(
      color: Colors.purple.shade400,
      borderRadius: BorderRadius.circular(50),
      ),
      child: const Padding(
      padding: EdgeInsets.symmetric(
          horizontal: 32.0,
          vertical: 16.0,
      ),
      child: Text(
          "Button",
          style: TextStyle(
          color: Colors.white,
          fontSize: 20,
          fontWeight: FontWeight.w500,
          ),
      ),
      ),
  ),
),
```


CHANGELOG.md
## 0.0.1

* Initial release.

## 0.0.2

* Fixed Issues

lib/with_loader.dart
library with_loader;

import 'package:flutter/material.dart';

class WithLoader<T> extends StatefulWidget {
  /// Function reference that you want to execute while your [child] is loading
  final Future<T> Function() future;

  /// Widget on which you want to having loading functionality
  final Widget child;

  /// Widget that replaces your [child] while your [child] is in loading state
  final Widget? loader;

  /// [onSuccess] or [onFailure] will execute based on this condition
  final bool Function(T) successCondition;

  /// Executes when [successCondition] is true
  final VoidCallback? onSuccess;

  /// Executes when [successCondition] is false
  final VoidCallback? onFailure;

  /// Alignment for [loader]. Defaults to [Alignment.center].
  final Alignment loaderAlignment;

  /// Opacity of [child] while it is in loading state. Defaults to 0.5.
  final double onLoadingOpacity;

  const WithLoader({
    Key? key,
    required this.future,
    required this.child,
    this.loader,
    required this.successCondition,
    this.onSuccess,
    this.onFailure,
    this.loaderAlignment = Alignment.center,
    this.onLoadingOpacity = 0.5,
  }) : super(key: key);

  @override
  _WithLoaderState<T> createState() => _WithLoaderState<T>();
}

class _WithLoaderState<T> extends State<WithLoader<T>>
    with TickerProviderStateMixin {
  bool isLoading = false;

  @override
  Widget build(BuildContext context) {
    return Stack(
      alignment: widget.loaderAlignment,
      children: [
        GestureDetector(
          onTap: () async {
            if (!isLoading) {
              setState(() {
                isLoading = true;
              });
              // Function executed
              var result = await widget.future();
              // Condition for success
              final isSuccess = widget.successCondition(result);

              if (isSuccess) {
                setState(() {
                  isLoading = false;
                });
                // If function is not null then execute
                widget.onSuccess?.call();
              } else {
                setState(() {
                  isLoading = false;
                });
                // If function is not null then execute
                widget.onFailure?.call();
              }
            }
          },
          child: Opacity(
            opacity: isLoading ? widget.onLoadingOpacity : 1.0,
            child: widget.child,
          ),
        ),
        if (isLoading)
          widget.loader != null
              ? widget.loader!
              : CircularProgressIndicator(
                  color: Theme.of(context).primaryColor,
                ),
      ],
    );
  }
}


example/lib/src/with_loader_example.dart
import 'package:example/src/backend_service.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:with_loader/with_loader.dart';

class WithLoaderExample extends StatelessWidget {
  const WithLoaderExample({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Center(
      child: WithLoader<String>(
        future: apiCall,
        successCondition: (result) => result == "Success",
        loaderAlignment: Alignment.center,
        onSuccess: () {
          if (kDebugMode) {
            print("Api Success");
          }
        },
        onFailure: () {
          if (kDebugMode) {
            print("Api Failure");
          }
        },
        child: Container(
          decoration: BoxDecoration(
            color: Colors.purple.shade400,
            borderRadius: BorderRadius.circular(50),
          ),
          child: const Padding(
            padding: EdgeInsets.symmetric(
              horizontal: 32.0,
              vertical: 16.0,
            ),
            child: Text(
              "Button",
              style: TextStyle(
                color: Colors.white,
                fontSize: 20,
                fontWeight: FontWeight.w500,
              ),
            ),
          ),
        ),
        loader: CircularProgressIndicator(
          color: Colors.purple.shade600,
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:example/backend_service.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:with_loader/with_loader.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'With Loader Example',
      theme: ThemeData(
        primarySwatch: Colors.purple,
      ),
      home: Scaffold(
          body: Center(
        child: WithLoader<String>(
          future: apiCall,
          successCondition: (result) => result == "Success",
          loaderAlignment: Alignment.center,
          onSuccess: () {
            if (kDebugMode) {
              print("Api Success");
            }
          },
          onFailure: () {
            if (kDebugMode) {
              print("Api Failure");
            }
          },
          child: Container(
            decoration: BoxDecoration(
              color: Colors.purple.shade400,
              borderRadius: BorderRadius.circular(50),
            ),
            child: const Padding(
              padding: EdgeInsets.symmetric(
                horizontal: 32.0,
                vertical: 16.0,
              ),
              child: Text(
                "Button",
                style: TextStyle(
                  color: Colors.white,
                  fontSize: 20,
                  fontWeight: FontWeight.w500,
                ),
              ),
            ),
          ),
          loader: CircularProgressIndicator(
            color: Colors.purple.shade600,
          ),
        ),
      )),
    );
  }
}
